cmd/juju/commands/upgrademodel.go

```diff
@@ -396,14 +403,328 @@ func (c *upgradeJujuCommand) Run(ctx *cmd.Context) (err error) {
+	modelUpgrader, err := c.getModelUpgraderAPI()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer modelUpgrader.Close()
+
+	if modelUpgrader.BestAPIVersion() == 0 {
+		// TODO(juju3) - remove
+		return c.upgradeModelLegacy(ctx, modelType, c.timeout)
+	}
+	return c.upgradeModel(ctx, modelUpgrader, c.timeout, modelType)
+}
+
+func uploadTools(
+	modelUpgrader ModelUpgraderAPI, buildAgent bool, agentVersion version.Number, dryRun bool,
+) (targetVersion version.Number, err error) {
+	builtTools, err := sync.BuildAgentTarball(
+		buildAgent, "upgrade",
+		func(builtVersion version.Number) version.Number {
+			builtVersion.Build++
+			if agentVersion.Build >= builtVersion.Build {
+				builtVersion.Build = agentVersion.Build + 1
+			}
+			targetVersion = builtVersion
+			return builtVersion
+		},
+	)
+	if err != nil {
+		return targetVersion, errors.Trace(err)
+	}
+	defer os.RemoveAll(builtTools.Dir)
+
+	if dryRun {
+		logger.Debugf("dryrun, skipping upload tools")
+		return targetVersion, nil
+	}
+
+	uploadToolsVersion := builtTools.Version
+	uploadToolsVersion.Number = targetVersion
+
+	toolsPath := path.Join(builtTools.Dir, builtTools.StorageName)
+	logger.Infof("uploading agent binary %v (%dkB) to Juju controller", targetVersion, (builtTools.Size+512)/1024)
+	f, err := os.Open(toolsPath)
+	if err != nil {
+		return targetVersion, errors.Trace(err)
+	}
+	defer f.Close()
+
+	_, err = modelUpgrader.UploadTools(f, uploadToolsVersion)
+	if err != nil {
+		return targetVersion, errors.Trace(err)
+	}
+	return targetVersion, nil
+}
+
+func (c *upgradeJujuCommand) upgradeWithTargetVersion(
+	ctx *cmd.Context, modelUpgrader ModelUpgraderAPI, isControllerModel, dryRun bool,
+	modelType model.ModelType, targetVersion, agentVersion version.Number,
+) (chosenVersion version.Number, err error) {
+
+	chosenVersion = targetVersion
+	// juju upgrade-controller --agent-version 3.x.x
+	_, err = c.notifyControllerUpgrade(ctx, modelUpgrader, targetVersion, dryRun)
+	if err == nil {
+		// All good!
+		// Upgraded to the provided target version.
+		logger.Debugf("upgraded to the provided target version %q", targetVersion)
+		return chosenVersion, nil
+	}
+	if !errors.Is(err, errors.NotFound) {
+		return chosenVersion, err
+	}
+
+	// If target version is the current local binary version, then try to upload.
+	canImplicitUpload := CheckCanImplicitUpload(
+		modelType, isOfficialClient(), jujuversion.Current, agentVersion,
+	)
+	if !canImplicitUpload || !isControllerModel {
+		// expecting to upload a local binary but we are not allowed to upload, so pretend there
+		// is no more recent version available.
+		logger.Debugf("no available binary found, and we are not allowed to upload, err %v", err)
+		return chosenVersion, errUpToDate
+	}
+
+	if targetVersion.Compare(jujuversion.Current.ToPatch()) != 0 {
+		logger.Warningf(
+			"try again with --agent-version=%s if you want to upload the local binary",
```

Comment:
this is the jujud from the snap right?
we could reword the message to be better if that's the case
"try again with --agent-version=%s if you want to upgrade using the packaged jujud from the snap"